gen.go

```diff
@@ -528,6 +533,97 @@ func emitCborMarshalSliceField(w io.Writer, f Field) error {
 		return xerrors.Errorf("Byte array in field {{ .Name }} was too long")
 	}
 
+	{{ if .PreserveNil }}
+	if {{ .Name }} == nil {
+		_, err := w.Write(cbg.CborNull)
+		if err != nil {
+			return err
+		}
+	} else {
+	{{ end }}
+		{{ MajorType "cw" "cbg.MajByteString" (print "len(" .Name ")" ) }}
+
+		if _, err := cw.Write({{ .Name }}[:]); err != nil {
+			return err
+		}
+	{{ if .PreserveNil }}
+	}
+	{{ end }}
+`)
+	}
+
+	var pointer bool
+	if e.Kind() == reflect.Ptr {
+		e = e.Elem()
+		pointer = true
+	}
+
+	err := doTemplate(w, f, `
+	if len({{ .Name }}) > {{ MaxLen .MaxLen "cbg.MaxLength" }} {
+		return xerrors.Errorf("Slice value in field {{ .Name }} was too long")
+	}
+
+	{{ if .PreserveNil }}
```

Comment:
Given that we now have this tag, do we need to duplicate the array code? We'll only have a compile error if the user intentionally specifies "preservenil" on an array, which, IMO, we should have anyways.

Reply:
hrm... my concern is that the template conditionals are getting overly complicated. We have the .IsArray check and the .PreserveNil check in the same template and i think theyre overlapping enough to make it really hard to read

Reply:
Ah, I see. This code is just getting a bit long. But you're right.